### DISCUSSION TITLE
Logs of connections

### DISCUSSION THREAD

**Discussion by jpergentino** [wg-easy/wg-easy#732] [Q&A] [answered]
Hi all,
is there a way to get the history of connected clients?

I know that I can see the list of connected devices on UI, but I would like to audit the connections.

Thank you in advance.

**Comment by peterlewis** [comment]
From [u/whythehellnote](https://www.reddit.com/user/whythehellnote/) on [Reddit](https://www.reddit.com/r/WireGuard/comments/m08oan/comment/gq6zpxv/):

> Wireguard is connectionless, so "when a connection is made" doesn't make much sense in the context of wireguard. Information wireguard might have is "last packet received", "last packet sent", "last keepalive", "last handshake". (A handshake is used to exchange and update keys to keep the channel secure, but that doesn't count as a connection in the traditional  sense of the word)
>
>If my phone is sat on my wifi with a wireguard tunnel up pinging merrily away, and then I walk out of the house, the wifi drops, and the phone switches to 4g, the ping continues, but the IP my encrypted wireguard packets come from changes. Is that a new connection in your view?
>
>What if I get in my car and drive through a tunnel, losing signal in the tunnel, but regaining it afterwards. New connection?
>
>How about I reboot my phone, so I stop getting traffic from the phone for a couple of minutes. Phone comes back, and traffic comes back. Is that a new connection?
>
>What if I get in a plane and fly across an ocean?
>
>If I ping a server once a second, I'm emitting a packet every 1000ms, and getting a response 2ms later. The rest of the time, am I connected or not? When does my connection "time out"? After a second? After a minute? After 15 minutes?
>
>"currently active" only really means a packet is currently being forwarded through the kernel, or it's an arbitary time since the last packet or key exchange (a timeout). There's no concept of logging off in wireguard, so the only way for a "connection" to end would be through a timeout since the last traffic.

**Comment by eandersons** [comment, marked as answer]
You could try [Pro Custodibus](https://www.procustodibus.com/features/) [Community Edition](https://www.procustodibus.com/editions/#ce) which seems to be doing what you are looking for:

> Usage monitoring
> 
>   - [..]
>   - Compare with historical access patterns
>   - [..]
>
> [..]
>
> Audit logging
>
>   - Connection history and traffic logging of each monitored host
>   - [..]

You will have to figure out how to use it with WireGuard Easy, but it should be fairly easy - you just have to add the needed Pro Custodibus containers (for [agent](https://docs.procustodibus.com/guide/agents/container/) and [API, database and app](https://docs.procustodibus.com/guide/onpremises/install/#docker)) that are using the same volume where `wg0.conf` is stored as WireGuard Easy container.

Disclaimer: I have not used Pro Custodibus, so getting it to work with WireGuard Easy might be more complex than what I wrote above.
At this point it actually could be easier to replace WireGuard Easy with Pro Custodibus entirely, just remember to back up `wg0.conf` to keep configuration of existing peers.
Actually today I could give Pro Custodibus a try to see how easy it is to set it up, how it looks and if it can work together with WireGuard Easy.

You could also try to [build your own WireGuard logging system](https://www.procustodibus.com/blog/2021/01/how-to-monitor-wireguard-activity/#build-your-own-analytics).

**Reply by eandersons** [reply]
> Disclaimer: I have not used Pro Custodibus, so getting it to work with WireGuard Easy might be more complex than what I wrote above.

Indeed this was not as easy as I imagined, but I managed to get Pro Custodibus work with WireGuard Easy.

![image](https://github.com/wg-easy/wg-easy/assets/9976861/d72176ca-a23d-46f6-9892-e304f6eb425a)

As it can be seen in the screenshot above, endpoint's activity can be viewed for up to 3 months. If data should be persisted, [a broker](https://docs.procustodibus.com/guide/integrations/brokers/) should be added to push events either to syslog or JSON file which in turn then can be utilized by some aggregation tool or [a webhook](https://docs.procustodibus.com/guide/integrations/webhooks/) should be added to send events to some web service.

One notable downside is that in Pro Custodibus peers are displayed by their public key, so they should be manually renamed. Name of peers can be observed in WireGuard Easy's `wg0.conf`.

If anyone is interested, feel free to ask, I will share steps that are needed to get WireGuard Easy work together with Pro Custodibus.

> At this point it actually could be easier to replace WireGuard Easy with Pro Custodibus entirely [..].

I was not completely right about this as well. While Pro Custodibus can be used to manage WireGuard clients and replace WireGuard Easy, it is targeted for a different use case and it certainly is not easy as user interface seems a bit confusing at first and more actions are needed to configure a new WireGuard client as opposed to WireGuard Easy.

**Reply by dennisdre** [reply]
> If anyone is interested, feel free to ask, I will share steps that are needed to get WireGuard Easy work together with Pro Custodibus.

I would be very much interested in the steps that you have taken to get this to work!
Thanks!!

**Reply by eandersons** [reply]
Hi, @dennisdre,

I will try to get to it (I will probably create a little repository for this) when I will have free time.

**Reply by dennisdre** [reply]
Thanks! That would be awesome!!

**Reply by eandersons** [reply]
I finally managed to get back to this, sorry for a long delay!

Basically a custom WireGuard Easy image is needed that has the Pro Custodibus agent baked in and a Pro Custodibus instance.

First [a Pro Custodibus instance should be set up](https://docs.procustodibus.com/guide/onpremises/install/#docker) (note that it also has a WireGuard container, so a different port should be mapped to the host) and [a new host should be added](https://docs.procustodibus.com/guide/hosts/add/) to it.

Then the custom WireGuard Easy image with a baked in Pro Custodibus agent should be built.

`wg-easy-pro-custodibus-agent.Dockerfile`:

```Dockerfile
ARG pro_custodibus_agent_version=1.4.5
ARG wireguard_easy_version=11

FROM procustodibus/agent:${pro_custodibus_agent_version} as pro-custodibus-agent


FROM ghcr.io/wg-easy/wg-easy:${wireguard_easy_version}

ARG pro_custodibus_agent_version=1.4.5
ARG pro_custodibus_prefix=procustodibus-agent
ARG pro_custodibus_dir="${pro_custodibus_prefix}-${pro_custodibus_agent_version}"
ARG pro_custodibus_archive="${pro_custodibus_dir}.tar.gz"

COPY --from=pro-custodibus-agent \
  ["/opt/venvs/procustodibus-agent", "/opt/venvs/procustodibus-agent"]
COPY ["./image-files", "/"]

RUN apk update && \
  apk add --no-cache libsodium openrc python3 && \
  apk cache clean && \
  wget https://ad.custodib.us/agents/"${pro_custodibus_archive}" && \
  tar --extract --file "${pro_custodibus_archive}" && \
  rm "${pro_custodibus_archive}" && \
  mv "${pro_custodibus_dir}" /opt/"${pro_custodibus_prefix}" && \
  chmod -R +x /etc/init.d/ && \
  # https://git.sr.ht/~arx10/procustodibus-agent/tree/main/item/docker/wireguard.dockerfile
  sed -i 's/^\(tty\d\:\:\)/#\1/' /etc/inittab && \
  sed -i \
    -e 's/^#\?rc_env_allow=.*/rc_env_allow="\*"/' \
    -e 's/^#\?rc_sys=.*/rc_sys="docker"/' \
    /etc/rc.conf && \
  sed -i \
    -e 's/VSERVER/DOCKER/' \
    -e 's/checkpath -d "$RC_SVCDIR"/mkdir "$RC_SVCDIR"/' \
    /lib/rc/sh/init.sh && \
  rm /etc/init.d/hwdrivers /etc/init.d/machine-id && \
  sed -i \
    's/cmd sysctl -q \(.*\?\)=\(.*\)/[[ "$(sysctl -n \1)" != "\2" ]] \&\& \0/' \
    /usr/bin/wg-quick && \
  # https://git.sr.ht/~arx10/procustodibus-agent/tree/main/item/docker/agent.dockerfile
  for file in /opt/venvs/"${pro_custodibus_prefix}"/bin/procustodibus-*; \
    do ln -fs "${file}" /usr/local/bin/; \
  done && \
  rc-update add procustodibus-agent default && \
  rc-update add wg-easy default

CMD ["/sbin/init"]

```

`image-files/etc/init.d/procustodibus-agent` (it is copied into the custom image):

```shell
#!/sbin/openrc-run

command=/usr/local/bin/procustodibus-agent
command_args="--loop=120 --verbosity=INFO"
command_background=true
name="Pro Custodibus Agent"
error_log=/dev/stderr
output_log=/dev/stdout
pidfile=/var/run/procustodibus-agent.pid

depend() {
  need localmount
  need net
  after wg-easy
}

```

`image-files/etc/init.d/wg-easy` (it is copied into the custom image):

```shell
#!/sbin/openrc-run

command=/usr/local/bin/node
command_args=/app/server.js
command_background=true
name="WireGuard Easy"
error_log=/dev/stderr
output_log=/dev/stdout
pidfile=/var/run/WireGuard-Easy.pid

depend() {
  need localmount
  need net
}

```

`.env` (provides value for variables that are used in `compose[.override].yaml`):

```shell
PRO_CUSTODIBUS_AGENT_VERSION=1.4.5
WIREGUARD_EASY_VERSION=11

```

`compose.override.yaml` (the relevant parts are `build` and `image`):

```yaml
services:
  wg-easy:
    build:
      args:
        pro_custodibus_agent_version: ${PRO_CUSTODIBUS_AGENT_VERSION:?}
        wireguard_easy_version: ${WIREGUARD_EASY_VERSION:?}
      dockerfile: wg-easy-pro-custodibus-agent.Dockerfile
    environment:
      WG_HOST: wireguard.example.org
      PASSWORD: strongpassword
      WG_DEFAULT_ADDRESS: 10.0.0.x
    image: wg-easy-pro-custodibus-agent:${WIREGUARD_EASY_VERSION}-${PRO_CUSTODIBUS_AGENT_VERSION}
    ports:
      - 0.0.0.0:51820:51820/udp
      - 127.0.0.1:51821:51821/tcp
    volumes:
      - wireguard:/etc/wireguard
      # Might not be needed. Unfortunately I do not remember why this mount was added.
      - pro-custodibus-cgroup:/sys/fs/cgroup

volumes:
  pro-custodibus-cgroup:
  wireguard:

```

The previously downloaded `procustodibus.conf` and `procustodibus-setup.conf` should be added to the volume or directory that will be mounted into the WireGuard Easy container at `/etc/wireguard`.

When all the files above have been added/modified, then image should be built and WireGuard Easy container started:

```shell
docker compose build
docker compose up --detach
```

I am not completely sure, but `wg0.conf` should be synchronized from the WireGuard Easy container to the Pro Custodibus stack's WireGuard container.

I hope that I have not forgotten anything important, it has been a while since I did this.

**Comment by jpergentino** [comment]
It would be good to have it embedded in the docker image, configurable of
course for the ones that don't want this.

**Comment by dennisdre** [comment]
> I finally managed to get back to this, sorry for a long delay!
> 
Looks great!! Thanks for this! i wil test it in a couple of days!
Thanks again for your effort!
